### PR TITLE
Fix metadata constructor inference by phpstan

### LINKED ISSUE #8709: The class ClassMetada is not correctly inferring the constructor of ClassMetadaInfo
Hi

Currently
- `new ClassMetadataInfo(\DateTime::class)` is inferred as `ClassMetadataInfo<\DateTime>` by both psalm and phpstan
- `new ClassMetadata(\DateTime::class)` is inferred as `ClassMetadata<\DateTime>` by psalm but as `ClassMetadata<object>` by phpstan

This is related to this phpstan issue https://github.com/phpstan/phpstan/issues/5057
I don't know how and when the issue will be fixed so one way to fix it temporary would be to repeat the constructor:
```
        /**
	 * @psalm-param class-string<T> $class
	 **/
	public function __construct(string $class)
	{
		parent::__construct($class);
	}
```

Since you're using phpstan in this project, you might end with this issue while bumping the phpstan level.
Do you prefer I create a PR here or in https://github.com/phpstan/phpstan-doctrine ?
(The second solution won't fix the issue here since you're not using the library here)

### PR DESCRIPTION
**Pull request by VincentLanglet**
Close https://github.com/doctrine/orm/issues/8709

### CODE REVIEW

File: phpstan.neon [hunk 1]
```diff
@@ -5,6 +5,7 @@ parameters:
     level: 5
     paths:
         - %currentWorkingDirectory%/lib
+        - %currentWorkingDirectory%/tests/Doctrine/StaticAnalysis
```
Comment (greg0ire) [inline phpstan.neon]: Let's get rid of `%currentWorkingDirectory%` instead? It does not bring any benefit.

**Review by greg0ire — COMMENTED** [review]
Do any errors go away if you regenerate Psalm's baseline? I think PHPStan would complain if it was the case for its baseline.

**Comment by VincentLanglet** [comment]
> Do any errors go away if you regenerate Psalm's baseline? I think PHPStan would complain if it was the case for its baseline.

PHPStan does complain.
No error will be removed from Psalm baseline since Psalm already correctly inferred the constructor.

File: phpstan.neon (hunk 1 above)
Comment (VincentLanglet) [reply phpstan.neon]: Done

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
@orklah please review

**Review by ostrolucky — APPROVED** [review]

**Comment by orklah** [comment]
It's a shame to fix that in Doctrine's code but there's not much else to do...

I'm quite puzzled by the tests though. I wonder if it's a good move to add tests to check that Psalm and PHPStan understand an arbitrary piece of code. I'm afraid future errors in those tests could lead to version stagnation for those tools while providing little value for the user (they will still upgrade their versions and their analysis will still be broken).

I'd much rather see PHPStan and Psalm add doctrine to their own CI instead (Psalm already does this for a few libraries that relies on static analysis)

**Comment by greg0ire** [comment]
About the tests, this approach was suggested by both beberlei and Ocramius, see https://github.com/doctrine/orm/issues/8709#issuecomment-847303948 . IMO it's good to test that the annotations we write are correct and don't break downstream projects.

As for the deps stagnation, I'm not too worried about it, I think we will regularly upgrade just like we do now, to get bugfixes.

**Comment by greg0ire** [comment]
Thanks @VincentLanglet !

**Pull request merged**